### PR TITLE
Caisson ID Check integration

### PR DESCRIPTION
**Pull request by ken-little**
### Description
Caisson ID Check rule.  Caisson validates US driver's licenses and international passports in real time.

### Testing

- Install the Caisson ID Check rule
- Sign up for an account at [Caisson](https://www.caisson.com)
- Go to your [developer tab](https://www.caisson.com/developer) to fetch your API keys 
- In Auth0, set the following rule configurations
  - CAISSON_PUBLIC_KEY
  - CAISSON_PRIVATE_KEY
  - CAISSON_ON_REGISTRATION
    - "true" (require ID Check on registration)
    - "false" / empty
  - CAISSON_LOGIN_FREQUENCY_DAYS
    - "-1" (always)
    - "0" / empty (never)
    - "[uint]" (min days between ID Checks on login)
  - CAISSON_DEBUG
    - "true" (prints errors to console)
    - "false" / empty

Go through your test app's login flow as usual.  You'll be taken through the ID Check process.  Check for caisson attributes on your user's app_metadata.  You should see: 
  - "idcheck_url" (link to your [Caisson Dashboard](https://www.caisson.com/dashboard) with full ID Check results)
  - "last_check" (timestamp of last ID Check)
  - "status" (Pass / Flagged status for the last ID Check)
  - "count" (number of completed ID Checks) 

### Checklist

- [x] All active GitHub checks for tests, formatting, and security are passing

### CODE REVIEW

File: src/rules/caisson-id-check.js [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+      public_key: caissonConf.CAISSON_PUBLIC_KEY,
+      private_key: caissonConf.CAISSON_PRIVATE_KEY,
+    },
+    debug:
+      caissonConf.CAISSON_DEBUG &&
+      caissonConf.CAISSON_DEBUG.toLowerCase() === "true"
+        ? true
+        : false,
+    idCheckFlags: {
+      login_frequency_days: parseInt(caissonConf.CAISSON_LOGIN_FREQUENCY_DAYS),
```
Comment (joshcanhelp) [inline src/rules/caisson-id-check.js:28]:
```suggestion
      login_frequency_days: parseInt(caissonConf.CAISSON_LOGIN_FREQUENCY_DAYS, 10),
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt

File: src/rules/caisson-id-check.js [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+  }
+
+  /**
+   * Creates Caisson specific session token and sets redirect url.
+   */
+  function setIDCheckRedirect() {
+    const token = manager.util.createSessionToken({
+      public_key: manager.creds.public_key,
+      host: context.request.hostname,
+      user_id: user.user_id,
```
Comment (joshcanhelp) [inline src/rules/caisson-id-check.js:65]: The user ID is already added as `sub` in there, just FYI

File: src/rules/caisson-id-check.js [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    //throws if redirects aren't allowed here.
+    manager.util.doRedirect(`${manager.caissonHosts.idcheck}/auth0`, token); //throws
+  }
+
+  /**
+   * Swaps the temp Caisson exchange token for an ID Check key.
+   * https://www.caisson.com/docs/reference/api/#exchange-check-token-for-check-id
+   * @param {string} t
+   */
+  async function exchangeToken(t) {
```
Comment (joshcanhelp) [inline src/rules/caisson-id-check.js:77]: Is this parameter used? Looks like you're also looking for the query param

File: src/rules/caisson-id-check.js [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+      user_id: user.user_id,
+    });
+
+    //throws if redirects aren't allowed here.
+    manager.util.doRedirect(`${manager.caissonHosts.idcheck}/auth0`, token); //throws
+  }
+
+  /**
+   * Swaps the temp Caisson exchange token for an ID Check key.
+   * https://www.caisson.com/docs/reference/api/#exchange-check-token-for-check-id
```
Comment (joshcanhelp) [inline src/rules/caisson-id-check.js:74]: 👍

File: src/rules/caisson-id-check.js [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+          headers: {
+            Authorization: `Caisson ${manager.creds.private_key}`,
+            "X-Caisson-CheckID": check_id,
+          },
+        }
+      );
+
+      if (resp.data.error) {
+        throw new Error(
+          "error in Caisson ID Check: " + JSON.stringify(resp.data)
```
Comment (joshcanhelp) [inline src/rules/caisson-id-check.js:120]:
```suggestion
          "Error in Caisson ID Check: " + JSON.stringify(resp.data)
```

File: src/rules/caisson-id-check.js [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+            Authorization: `Caisson ${manager.creds.private_key}`,
+          },
+        }
+      );
+
+      return resp.data.check_id;
+    } catch (error) {
+      let err = error;
+      if (err.response && err.response.status === 401) {
+        err = new Error(
```
Comment (joshcanhelp) [inline src/rules/caisson-id-check.js:93]:
```suggestion
        err = new UnauthorizedError(
```

Maybe?

File: src/rules/caisson-id-check.js [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+      const check_id = await exchangeToken(manager.util.queryParams.t);
+      const results = await idCheckResults(check_id);
+      await updateUser(results);
+    } catch (err) {
+      dLog(err);
+      return callback(err);
+    }
+
+    return callback(null, user, context);
+  } else {
```
Comment (joshcanhelp) [inline src/rules/caisson-id-check.js:219]: No `else` needed

File: src/rules/caisson-id-check.js [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+    /**
+     * Perform ID Checks when appropriate
+     */
+    user.app_metadata = user.app_metadata || {};
```
Comment (joshcanhelp) [inline src/rules/caisson-id-check.js:223]: Duplicated in `updateUser()`, recommend doing this once at the top.

**Review by joshcanhelp — CHANGES_REQUESTED** [review]
This is really well-done, nice work here! A few callouts and a handful of changes needed.

File: src/rules/caisson-id-check.js [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+      }
+
+      let results = {
+        check_id: resp.data.check_id,
+        auth0_id: resp.data.customer_id,
+        timestamp: resp.data.checked_on,
+        status:
+          resp.data.confidence.document === "high" &&
+          resp.data.confidence.face === "high"
+            ? "passed"
```
Comment (joshcanhelp) [inline src/rules/caisson-id-check.js:131]: Same as above ... lets move this to a single line to avoid the error.

**Review by joshcanhelp — CHANGES_REQUESTED** [review]
Tested end-to-end ... works great! A few more things that jump out at me.

File: src/rules/caisson-id-check.js (hunk 2 above)
Comment (ken-little) [reply src/rules/caisson-id-check.js:65]: moved to using "sub"

File: src/rules/caisson-id-check.js (hunk 6 above)
Comment (ken-little) [reply src/rules/caisson-id-check.js:93]: Moved both 401's to UnauthorizedError

**Review by joshcanhelp — APPROVED** [review]

File: src/rules/caisson-id-check.js [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+      throw err;
+    }
+  }
+
+  /**
+   * ID Check is done, handle results.
+   */
+  if (manager.util.isRedirectCallback) {
+    //is it our redirect?
+    if (!manager.util.queryParams.caisson_flow) {
```
Comment (joshcanhelp) [inline src/rules/caisson-id-check.js:194]: Note from security: let's check the value here as well.

**Review by joshcanhelp — CHANGES_REQUESTED** [review]
One minor change from the security review and we're good to go!

**Review by joshcanhelp — APPROVED** [review]

**Pull request merged**